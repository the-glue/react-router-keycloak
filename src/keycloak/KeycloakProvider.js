import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import KeycloakContext from './KeycloakContext';
import { updateToken, isAuthenticated } from './keycloak';

const propTypes = {
  loginPath: PropTypes.string.isRequired,
  logoutPath: PropTypes.string.isRequired,
  onRefresh: PropTypes.func.isRequired,
  minValidity: PropTypes.number,
  refreshRate: PropTypes.number,
  children: PropTypes.any
};
const defaultProps = {
  minValidity: 30,
  refreshRate: 10
};

export default function KeycloakProvider(props) {
  const { loginPath, logoutPath, onRefresh, minValidity, refreshRate, children } = props;
  useEffect(() => {
    const refreshInterval = setInterval(() => {
      if (isAuthenticated()) {
        updateToken(onRefresh, minValidity);
      }
    }, refreshRate * 1000);
    return () => {
      clearInterval(refreshInterval);
    };
  });
  return (
    <KeycloakContext.Provider value={{ loginPath, logoutPath, onRefresh, minValidity, refreshRate }}>
      {children}
    </KeycloakContext.Provider>
  );
}
KeycloakProvider.propTypes = propTypes;
KeycloakProvider.defaultProps = defaultProps;
