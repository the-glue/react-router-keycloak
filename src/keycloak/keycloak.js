import Keycloak from 'keycloak-js';

// Setup client
export let keycloak;

/**
 * Instantiate a keycloak client
 * @param {string} keycloakUrl
 * @param {string|number} realm
 * @param {string|number} clientId
 */
export function configureKeycloak(keycloakUrl, realm, clientId) {
  if (!keycloakUrl) {
    throw new Error('There is no Keycloak url configured');
  }
  if (!realm) {
    throw new Error('There is no realm configured');
  }
  if (!clientId) {
    throw new Error('There is no client id configured');
  }
  if (!keycloak) {
    keycloak = new Keycloak({
      url: `${keycloakUrl}/auth/`,
      realm: realm,
      clientId: clientId
    });
  }
}

/**
 * Get the keycloak instance
 * @returns {Keycloak}
 */
export function getKeycloak() {
  if (!keycloak) {
    throw new Error('Please configure keycloak first');
  }
  return keycloak;
}

/**
 * Calls the keycloak updateToken then calls the onRefresh callback on success if the min validity is reached
 * @param {Function} onRefresh Callback function called on refresh success
 * @param {Number} [minValidity=30] If the token expires within `minValidity` seconds, the token is refreshed.
 */
export const updateToken = (onRefresh, minValidity = 30) => {
  const keycloak = getKeycloak();
  keycloak.updateToken(minValidity).then(refreshed => {
    if (refreshed) {
      onRefresh(keycloak.token);
    }
  });
};

/**
 * Check if we are authenticated or not on keycloak
 * @returns {boolean}
 */
export const isAuthenticated = () => {
  const keycloak = getKeycloak();
  return keycloak.authenticated;
};
