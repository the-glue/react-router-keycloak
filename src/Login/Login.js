import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { getKeycloak } from '../keycloak/keycloak';

class Login extends React.Component {
  static propTypes = {
    onSuccess: PropTypes.func,
    onFailure: PropTypes.func,
    redirectTo: PropTypes.string.isRequired,
    location: PropTypes.shape({
      state: PropTypes.shape({
        redirectTo: PropTypes.string
      })
    }),
    children: PropTypes.any
  };
  static defaultProps = {
    onSuccess: () => {},
    onFailure: e => console.error(e)
  };

  state = {
    authenticated: false
  };

  componentDidMount() {
    this.logIn();
  }

  logIn = () => {
    const keycloak = getKeycloak();
    keycloak
      .init()
      .success(authenticated => {
        if (authenticated) {
          // Remove the last redirect path from the session storage
          window.sessionStorage.removeItem('keycloak-react-router:redirectTo');
          // Update the state to re-render so it will redirect to the previous private route
          this.setState({ authenticated });
          // Call the onSuccess callback with the provided keycloak token
          this.props.onSuccess(keycloak.token);
        } else {
          try {
            const { location } = this.props;
            if (location.state && location.state.redirectTo) {
              // Store the current path in session storage before leaving the application to log in on keycloak server
              window.sessionStorage.setItem('keycloak-react-router:redirectTo', location.state.redirectTo);
            }
            // Redirect to keycloak login page
            keycloak.login();
          } catch (e) {
            this.props.onFailure(e);
          }
        }
      })
      .error(() => {
        this.props.onFailure('There was an error with initializing keycloak, please check your credentials');
      });
  };

  render() {
    const { children, redirectTo } = this.props;
    if (!this.state.authenticated) {
      // display default loading or the one provided as children
      if (children) {
        return children;
      }

      return <div>Loading...</div>;
    }
    // redirect to the assigned path in the session storage or fallback to the one provided as props
    return <Redirect to={window.sessionStorage.getItem('keycloak-react-router:redirectTo') || redirectTo} />;
  }
}

export default Login;
