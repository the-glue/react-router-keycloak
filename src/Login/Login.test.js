import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Login from './Login';
import { configureKeycloak } from '../keycloak/keycloak';

describe('Login', () => {
  beforeAll(() => {
    configureKeycloak('dummy url', 'dummy realm', 'dummy id');
  });

  it('renders without crashing given the required props', () => {
    const props = {
      redirectTo: '/',
      onSuccess: jest.fn(),
      onFailure: jest.fn()
    };
    const wrapper = renderer
      .create(
        <MemoryRouter initialEntries={['/login']}>
          <Route path="/login" render={() => <Login {...props} />} />
        </MemoryRouter>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
