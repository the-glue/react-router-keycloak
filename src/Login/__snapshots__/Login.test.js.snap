// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without crashing given the required props 1`] = `
<div>
  Loading...
</div>
`;
